File: app/src/main/java/com/tkdev/weatherapp/Forecast/ForecastPresenterImpl.java
package com.tkdev.weatherapp.Forecast;

import com.tkdev.weatherapp.Forecast.model.ForecastRetrofit;
import com.tkdev.weatherapp.Common.MainContract;
import com.tkdev.weatherapp.Common.WeatherRetrofitImpl;

import retrofit2.Response;

import static com.tkdev.weatherapp.Common.PreferencesVariables.current_city;
import static com.tkdev.weatherapp.Common.PreferencesVariables.last_dt;

public class ForecastPresenterImpl implements MainContract.Presenter, ForecastPresenter, MainContract.APIForecastListener {

    private MainContract.View view;
    private ForecastRetrofit forecasts;
    private MainContract.Model model;

    public ForecastPresenterImpl(MainContract.View view) {
        this.view = view;
        this.model = new WeatherRetrofitImpl();
        this.forecasts = new ForecastRetrofit();
    }

    public ForecastRetrofit getForecasts() {
        return forecasts;
    }

    @Override
    public void onSuccessResponse(Response<ForecastRetrofit> response) {
        forecasts = response.body();
        view.update();
    }



    @Override
    public void onFailureResponse(String responseMessage) {
//        view.onFailUpdate(responseMessage);
    }


    @Override
    public void onDestroy() {
        view = null;
    }

    @Override
    public void onRequestWeather(String city) {
        if (current_city.equals("")) {
            current_city = "london";
            model.getWeather(this, current_city);
        } else if (
                last_dt <= System.currentTimeMillis() / 1000 - 600
                        ||
                        (!(current_city.equals(city))))
        {
            model.getWeather(this, city);
        } else {
            view.cancelUpdate();
        }
    }
}

File: app/src/main/java/com/tkdev/weatherapp/forecast/ForecastPresenterImpl.kt
package com.tkdev.weatherapp.forecast

import com.tkdev.weatherapp.common.MainContract
import com.tkdev.weatherapp.common.PreferencesVariables.Companion.current_city
import com.tkdev.weatherapp.common.PreferencesVariables.Companion.last_dt
import com.tkdev.weatherapp.common.WeatherRetrofitImpl
import com.tkdev.weatherapp.forecast.model.ForecastRetrofit
import retrofit2.Response

class ForecastPresenterImpl(private var view: MainContract.View) : MainContract.Presenter, ForecastPresenter, MainContract.APIForecastListener {

    private lateinit var forecasts: ForecastRetrofit
    private val model: MainContract.Model

    init {
        model = WeatherRetrofitImpl()
    }

    override fun onSuccessResponse(response: Response<ForecastRetrofit>) {
        forecasts = response.body()!!
        view.update()
    }

    override fun onFailureResponse(responseMessage: String) {
//        view.onFailUpdate(responseMessage);
    }

    override fun onDestroy() {

    }

    override fun onRequestWeather(city: String) {
        if (current_city == "") {
            current_city = "london"
            model.getWeather(this, current_city)
        } else if (last_dt <= System.currentTimeMillis() / 1000 - 600
                ||
                current_city != city) {
            model.getWeather(this, city)
        } else {
            view.cancelUpdate()
        }
    }

    override fun getForecasts(): ForecastRetrofit {
        return forecasts
    }
}